File: src/main/java/wraith/fabricaeexnihilo/json/recipe/CrucibleRecipeJson.java
package wraith.fabricaeexnihilo.json.recipe;

import com.google.gson.*;
import wraith.fabricaeexnihilo.api.crafting.ItemIngredient;
import wraith.fabricaeexnihilo.api.recipes.crucible.CrucibleRecipe;
import wraith.fabricaeexnihilo.json.BaseJson;
import wraith.fabricaeexnihilo.json.other.FluidVolumeJson;

import java.lang.reflect.Type;

public final class CrucibleRecipeJson extends BaseJson<CrucibleRecipe> {

    private CrucibleRecipeJson() {}

    public static final CrucibleRecipeJson INSTANCE = new CrucibleRecipeJson();

    @Override
    public CrucibleRecipe deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        var obj = json.getAsJsonObject();
        return new CrucibleRecipe(
                ItemIngredient.fromJson(obj.get("input"), context),
                FluidVolumeJson.INSTANCE.deserialize(obj.get("output"), FluidVolumeJson.INSTANCE.getTypeToken(), context)
        );
    }

    @Override
    public JsonElement serialize(CrucibleRecipe src, Type typeOfSrc, JsonSerializationContext context) {
        var obj = new JsonObject();
        obj.add("input", src.input().toJson(context));
        obj.add("output", FluidVolumeJson.INSTANCE.serialize(src.output(), FluidVolumeJson.INSTANCE.getTypeToken(), context));
        return obj;
    }

}

File: src/main/kotlin/exnihilofabrico/json/CrucibleRecipeJson.kt
package exnihilofabrico.json

import com.google.gson.*
import exnihilofabrico.api.crafting.ItemIngredient
import exnihilofabrico.api.recipes.crucible.CrucibleRecipe
import exnihilofabrico.json.FluidVolumeJson.TYPE_TOKEN
import java.lang.reflect.Type

object CrucibleRecipeJson: JsonDeserializer<CrucibleRecipe>, JsonSerializer<CrucibleRecipe> {
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): CrucibleRecipe {
        val obj = json.asJsonObject
        return CrucibleRecipe(
            ItemIngredient.fromJson(obj.get("input"), context),
            FluidVolumeJson.deserialize(obj.get("output"), TYPE_TOKEN, context)
        )
    }

    override fun serialize(src: CrucibleRecipe, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val obj = JsonObject()
        obj.add("input", src.input.toJson(context))
        obj.add("output", FluidVolumeJson.serialize(src.output, TYPE_TOKEN, context))
        return obj
    }

}